dca/methods_comparison.py

```diff
@@ -610,3 +610,344 @@ def fit_transform(self, X, n_components=None):
             n_components = self.n_components
         self.fit(X)
         return self.transform(X, n_components=n_components)
+
+
+class JPCA(object):
+    """ jPCA.
+
+    Parameters
+    ----------
+    n_components : even int (default=6)
+        Number of components to reduce X to.
+
+    mean_subtract: boolean (default=True)
+        Whether to subtract the cross-condition mean from each condition
+        before running jPCA.
+
+    Attributes
+    ----------
+    eigen_vecs_ : list
+        List of numpy eigenvectors from JPCA skew symmetric matrix sorted in
+        descending order by magnitude of eigenvalue.
+
+    eigen_vals_ : list
+        List of eigenvalues from JPCA skew symmetric matrix. The index
+        of each eigenvalue corresponds with the eigenvector in eigen_vecs_.
+
+    pca_ : sklearn.decomp.PCA object
+        PCA object used to transform X to X_red.
+
+    cross_condition_mean_ : ndarray (time, features)
+        Cross condition mean of X during fit.
+
+    """
+    def __init__(self, n_components=6, mean_subtract=True):
+        if n_components // 2 != n_components / 2:
+            raise ValueError("n_components must be even int")
+        self.n_components_ = n_components
+        self.mean_subtract_ = mean_subtract
+        self.eigen_vecs_ = None
+        self.eigen_vals_ = None
+        self.pca_ = None
+        self.cross_condition_mean_ = None
+
+    def fit(self, X):
+        """ Fit a jPCA model to X.
+
+        Parameters
+        ----------
+        X : ndarray (time, features) or (conditions, time, features)
+            Data to fit using jPCA model.
+
+        Returns
+        -------
+        self
+        """
+        if self.n_components_ > X.shape[1]:
+            raise ValueError("n_components is greater than number of features in X.")
+
+        if self.mean_subtract_:
+            self.cross_condition_mean_ = np.mean(X, axis=0)
+            X = self._cross_condition_subtract(X, self.cross_condition_mean_)
+
+        condition_indices = self._get_condition_indices(X)
+
+        X = np.vstack(X)
```

Comment:
I think this could be done more simply

```python
        X_flat = np.concatenate(X, axis=0)
        self.pca_ = PCA(n_components=self.n_components_)
        self.pca_.fit(X_flat)

        X_red = [self.pca_.transform(Xi) for Xi in X]
        dX = np.concatenate([np.diff(Xi, axis=1) for Xi in X_red], axis=0)
        X_prestate = np.concatenate([Xi[:-1] for Xi in X_red], axis=0)

        M_skew = self._fit_skew(X_prestate, dX)
        self.eigen_vals_, self.eigen_vecs_ = self._get_jpcs(M_skew)

        return self
```

I think, in general, if you can do something in a few lines of code and you only need it in one place, it's not worth it to create a separate function.

Reply:
Agreed, this is much nicer! Fixed to this